client/cli/src/runner.rs

```diff
@@ -173,29 +182,57 @@ impl<C: SubstrateCli> Runner<C> {
 
 	/// A helper function that runs a future with tokio and stops if the process receives the signal
 	/// `SIGTERM` or `SIGINT`.
-	pub fn run_subcommand<B, BC, BB>(self, subcommand: &Subcommand, builder: B) -> Result<()>
+	pub fn run_subcommand<BU, B, BA, IQ, CL>(self, subcommand: &Subcommand, builder: BU)
+		-> Result<()>
 	where
-		B: FnOnce(Configuration) -> sc_service::error::Result<BC>,
-		BC: ServiceBuilderCommand<Block = BB> + Unpin,
-		BB: sp_runtime::traits::Block + Debug,
-		<<<BB as BlockT>::Header as HeaderT>::Number as FromStr>::Err: Debug,
-		<BB as BlockT>::Hash: FromStr,
-		<<BB as BlockT>::Hash as FromStr>::Err: Debug,
+		BU: FnOnce(Configuration)
+			-> sc_service::error::Result<(Arc<CL>, Arc<BA>, IQ, TaskManager)>,
+		B: BlockT + for<'de> serde::Deserialize<'de>,
+		BA: sc_client_api::backend::Backend<B> + 'static,
+		IQ: sc_service::ImportQueue<B> + 'static,
+		<B as BlockT>::Hash: FromStr,
+		<<B as BlockT>::Hash as FromStr>::Err: Debug,
+		<<<B as BlockT>::Header as HeaderT>::Number as FromStr>::Err: Debug,
+		CL: UsageProvider<B> + BlockBackend<B> + StorageProvider<B, BA> + Send + Sync +
+		'static,
 	{
+		let chain_spec = self.config.chain_spec.cloned_box();
+		let network_config = self.config.network.clone();
+		let db_config = self.config.database.clone();
+
 		match subcommand {
-			Subcommand::BuildSpec(cmd) => cmd.run(self.config),
+			Subcommand::BuildSpec(cmd) => cmd.run(chain_spec, network_config),
 			Subcommand::ExportBlocks(cmd) => {
-				run_until_exit(self.tokio_runtime, cmd.run(self.config, builder))
+				let (client, _, _, task_manager) = builder(self.config)?;
+				run_until_exit(
+					self.tokio_runtime, cmd.run(client, db_config), Some(task_manager)
+				)
 			}
 			Subcommand::ImportBlocks(cmd) => {
-				run_until_exit(self.tokio_runtime, cmd.run(self.config, builder))
+				let (client, _, import_queue, task_manager) = builder(self.config)?;
+				run_until_exit(
+					self.tokio_runtime, cmd.run(client, import_queue), Some(task_manager)
+				)
 			}
 			Subcommand::CheckBlock(cmd) => {
-				run_until_exit(self.tokio_runtime, cmd.run(self.config, builder))
+				let (client, _, import_queue, task_manager) = builder(self.config)?;
+				run_until_exit(
+					self.tokio_runtime, cmd.run(client, import_queue), Some(task_manager)
+				)
 			}
-			Subcommand::Revert(cmd) => cmd.run(self.config, builder),
-			Subcommand::PurgeChain(cmd) => cmd.run(self.config),
-			Subcommand::ExportState(cmd) => cmd.run(self.config, builder),
+			Subcommand::Revert(cmd) => {
+				let (client, backend, _, mut task_manager) = builder(self.config)?;
+				cmd.run(client, backend)?;
+				task_manager.terminate();
```

Comment:
This is not ideal.... :thinking: 

I know why it is there and it makes sense but it is impossible to guess when you read the code for the first time.

Maybe it can stay that way for now as it is not easy to fix.

The issues are:
- The task manager should not even be there in the first place because those commands are not async. Therefore they don't use the tokio runtime. Therefore there is nothing to terminate and spawning a task should not even be allowed.
- BUT It would be bad to introduce another builder for synchronous operations. So probably we should assume all codes are async and put the blocking code in an async block (or function) for now.